### PR TITLE
Validation Review UI

### PR DESCRIPTION
**Pull request by Tyouxik**
# New features
- Threat model validation is visible in the bottom panel
image

- Validation step happens before the threat model can be sent for review

image

### CODE REVIEW

File: app/package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "react-konva-utils": "^0.2.0",
     "react-redux": "^7.2.6",
     "react-router-dom": "^6.4.1",
+    "react-window": "^1.8.10",
```
Comment (Tethik) [inline app/package.json:32]: Where is this used?

Comment (Tyouxik) [reply app/package.json:32]: It is not used. It was part of an attempt. I need to clean that up.

File: app/package.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
     "@fortawesome/free-brands-svg-icons": "^6.5.1",
     "@fortawesome/react-fontawesome": "^0.2.0",
     "@mui/icons-material": "^5.15.20",
-    "@mui/material": "^5.15.20",
+    "@mui/material": "^6.1.0",
```
[the remark this answers is not in the document]
Comment (Tyouxik) [reply app/package.json:16]: I did a test of the whole app locally, and nothing broke. I am gonna double check the MUI migration guide to make sure I didn't miss a spot

[the remark this answers is not in the document]
Comment (Tyouxik) [reply app/package.json:16]: I had to update Grid to Grid2, I did it successfully everywhere except in the Board.
If I change to Grid 2, it causes issues with Konva. I don't quite understand the behavior.

**Comment by Tethik** [comment]
I think the MUI upgrade broke the Team system lists on the Home page:
image

**Comment by Tethik** [comment]
image

**Comment by Tethik** [comment]
- [ ] Remember to clean up your `console.log` before merge - they should not be used in release.

**Comment by Tethik** [comment]
More general feedback:

## Quality Check Popup
* As a user, if I go to request a review and have never seen the Quality Check / Validation before, I don't know how I'm supposed to open the validation panel. There is nothing informing that it exists until you press "Request Review" and even so you don't know that you can open the panel. Users may end up opening that Quality Check popup over and over instead of checking the bottom panel.
* The popup could be taller/longer, currently the scroll kicks in a bit too early I think.
* Percentage shown is not explained what it means
  * It would be cool if we calculated this serverside and had it via an API, that way we could use it later for e.g. Target Specifications / Violations.
  * Also probably should calculate this score based on weights. Some rules are likely less important than others.

## Bottom Tab
* Model tab is not clear to a user what it's meant for
* I don't feel like the button that opens the validation panel belongs in the control panel, but I don't have any better suggestion so I think we can leave it there for now.
* Maybe in the future we should consider removing the tabs to simplify into a single "Validation" tab.


## Rules / Validation backend
* With the current StaticValidator ruleset, how would I add a rule that checks if a component has at least one threat/control? I don't see any `dal` reference passed which would allow to make those lookups in the database. Maybe it should allow for being more dynamic (I know, static is in the name).
* The `validateModel` route itself does not follow conventions from other api routes (I can probably refactor this quickly).

## KPIs / Metrics
* One potential way to track how useful this feature is could be to start measuring the review statuses - does this feature result in less meetings?
* Another thing to measure over time is the average quality score.

File: app/src/components/model/modals/ValidateBeforeReview.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+import CloseIcon from "@mui/icons-material/Close";
+import FactCheckIcon from "@mui/icons-material/FactCheck";
+import { useDispatch } from "react-redux";
+import { useModelID } from "../hooks/useModelID";
+import { useValidateQuery } from "../../../api/gram/validation";
+import { modalActions } from "../../../redux/modalSlice";
+import { MODALS } from "../../elements/modal/ModalManager";
+
+export function ValidateBeforeReview() {
```
Comment (Tethik) [inline app/src/components/model/modals/ValidateBeforeReview.js:24]: Consider renaming this to QualityCheck or similar instead, makes it easier to find.

**Review by Tethik — CHANGES_REQUESTED** [review]

File: api/src/resources/gram/v1/validation/validateModel.ts [hunk 4]
```diff
@@ -3,38 +3,30 @@
...
-      model = await dal.modelService.getById(modelId);
-
-      if (!model) {
-        return res.sendStatus(400);
-      }
-    } catch (error) {
-      return res.sendStatus(400);
+    if (dal.validationEngine.rules.length === 0) {
+      return res
+        .status(404)
```
Comment (Tethik) [inline api/src/resources/gram/v1/validation/validateModel.ts:12]: 404 here doesn't make sense. It should just return 200 with an empty list. Or are you handling this on the frontend?

File: api/src/resources/gram/v1/validation/validateModel.ts [hunk 5]
```diff
@@ -3,38 +3,30 @@
...
-        total: validationResults.length,
-        results: validationResults,
-      });
-    } catch (error) {
-      return res.sendStatus(400);
+    const validationResults = await dal.validationEngine.getResults(modelId);
+
+    if (validationResults.length === 0) {
+      return res
+        .status(404)
```
Comment (Tethik) [inline api/src/resources/gram/v1/validation/validateModel.ts:22]: Same here, odd use of 404

File: api/src/resources/gram/v1/validation/validateModel.ts (hunk 4 above)
Comment (Tethik) [reply api/src/resources/gram/v1/validation/validateModel.ts:12]: 400 errors usually denote client errors, which this is not.

**Review by Tethik — APPROVED** [review]

**Pull request merged**